Dockerfile

```diff
@@ -1,13 +1,25 @@
-FROM ubuntu:xenial AS builder
 # NOTE Customize the binary that is being built by providing `PROJECT` build-arg.
 # E.g. docker build --build-arg PROJECT=ethereum-poa-relay ...
 
+# This first stage prepares our dependencies to be built by `cargo-chef`.
+FROM rust as planner
+WORKDIR /parity-bridges-common
+RUN cargo install cargo-chef
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+# This second stage is where the dependencies actually get built.
+# The reason we split it from the first stage is so that the `COPY . .`
+# step doesn't blow our cache.
+FROM ubuntu:xenial AS cacher
+WORKDIR /parity-bridges-common
+
 # show backtraces
 ENV RUST_BACKTRACE 1
-
 ENV LAST_DEPS_UPDATE 2020-06-22
 
 # install tools and dependencies
+# TODO: Check if we need all these deps (e.g jq)
 RUN set -eux; \
```

Comment:
We should probably publish our own image with these things, which would speed things up as well: the `cacher` and `builder` phase would be based on the same image - now we have to keep both parts exactly the same to make sure that docker cache is utilized here, otherwise prepare the base image from scratch twice.

Reply:
Yeah, I think that a separate "base" build image is a good idea. I don't know how much time we'd save using it (since grabbing and building the dependencies is free after the first build), but it will definitely make for a more manageable Dockerfile